src/http/adapter.js

```diff
@@ -190,18 +202,20 @@ function wireupAction( state, resource, actionName, action, metadata, resources
 		resource.name, actionName, action.method, meta.url );
 
 	state.http.route( meta.url, action.method, function( req, res ) {
+		meta.getEnvelope( req, res );
+		var authCheck = meta.getPermissionCheck( req );
 		req._metricKey = meta.metricKey;
 		req._resource = resource.name;
 		req._action = actionName;
-		req._checkPermission = meta.getPermissionCheck( req );
+		req._checkPermission = authCheck;
 		var timer = meta.getTimer();
 		res.once( 'finish', function() {
 			timer.record( { name: 'HTTP_API_DURATION' } );
 		} );
 
-		if ( state.auth ) {
+		if ( authCheck ) {
 			meta.authAttempted();
-			checkPermissionFor( state, req.user, req.context, meta.alias )
+			authCheck()
 				.then( function onPermission( pass ) {
```

Comment:
I think that if this promise rejects the request hangs. Didn't dig into why, just something I noticed during testing.